# Capstone Project - Customer Churn Prediction: Data Wrangling
- Data Wrangling of Telco Customer Dataset

### Table of Contents
1. [Introduction](#in) <br>
2. [Data Overview](#do) <br>
3. [Data Management](#dm) <br>
4. [Data Description](#dd) <br>

# Introduction
<a id="in"> </a>

This IBM sample dataset is acquired from Kaggle: [Telco Customer Churn - Focused customer retention programs](#https://www.kaggle.com/blastchar/telco-customer-churn). The goal of this project is to analyze all relevant customer data and develop focused customer retention programs and predict behavior to retain customers.

Each row of the dataset represents a customer, each column contains customer’s attributes described on the column Metadata. The detailed info is described below,

- Customers who left within the last month – the column is called Churn
- Services that each customer has signed up for – phone, multiple lines, internet, online security, online backup, device protection, tech support, and streaming TV and movies
- Customer account information – how long they’ve been a customer, contract, payment method, paperless billing, monthly charges, and total charges
- Demographic info about customers – gender, age range, and if they have partners and dependents

We will walk through the data cleaning in the data wrangling section.

In [1]:
# Import packages
import pandas as pd
import numpy as np
import pandas_profiling as pf 
from datetime import datetime
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns   
import missingno as msno
from collections import Counter

In [2]:
# Load the data
telco = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv") 

# Data Overview
<a id="do"> </a>

In [3]:
def data_summary(data):
    """This is a function returns the basic summary of the dataset
    """
    print("__Data Source Summary__ \n")
    print("Rows: {}".format(data.shape[0]))
    print("Columns:{}\n".format(data.shape[1]))
    print("Features:{}\n".format(data.columns.tolist()))
    
    data.replace(" ", np.nan, inplace = True)
    print("Total Missing Value: {}\n".format(data.isnull().sum().values.sum()))
    print("Missing Value by Features: \n{}".format(data.isnull().sum()[data.isnull().sum()>0]))
    print("Percentage: {:.2f}%\n".format(telco.isnull().sum()[telco.isnull().sum()>0].values[0]/telco.shape[0]*100))
    
    print("Unique Value: \n{}".format(data.nunique()))
      

In [4]:
data_summary(telco)

__Data Source Summary__ 

Rows: 7043
Columns:21

Features:['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn']

Total Missing Value: 11

Missing Value by Features: 
TotalCharges    11
dtype: int64
Percentage: 0.16%

Unique Value: 
customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        653

# Data Management 
<a id="dm"> </a>

In [5]:
# drop missing values
telco.dropna(subset = ['TotalCharges'], how = 'any', inplace = True)

# reset the index
telco.reset_index(drop= True, inplace = True)

# drop the ID column
def drop(data, feat):
    """
    drop feature from data
    """
    data = data.drop(feat, axis=1)
    print('Dropping {}...'.format(feat))
    return data

telco = drop(telco, 'customerID')

# recode the senior citizen features
telco.SeniorCitizen = telco.SeniorCitizen.replace({1:"Yes", 0:"No" })


# get all the category fetures
cate_col = ['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService',
            'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 
            'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'Churn']

# get the numerical features
nume_col = ['MonthlyCharges', 'TotalCharges']

# Convert the objects into corresponding category
for feat in telco.columns:
    if feat in cate_col:
        telco[feat] = telco[feat].astype('category')
    if feat in nume_col:
        telco[feat] = telco[feat].astype('float')        
    


Dropping customerID...


# Data Description
<a id="dd"> </a>

In [6]:
# post check the data type
telco.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7032 entries, 0 to 7031
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   gender            7032 non-null   category
 1   SeniorCitizen     7032 non-null   category
 2   Partner           7032 non-null   category
 3   Dependents        7032 non-null   category
 4   tenure            7032 non-null   int64   
 5   PhoneService      7032 non-null   category
 6   MultipleLines     7032 non-null   category
 7   InternetService   7032 non-null   category
 8   OnlineSecurity    7032 non-null   category
 9   OnlineBackup      7032 non-null   category
 10  DeviceProtection  7032 non-null   category
 11  TechSupport       7032 non-null   category
 12  StreamingTV       7032 non-null   category
 13  StreamingMovies   7032 non-null   category
 14  Contract          7032 non-null   category
 15  PaperlessBilling  7032 non-null   category
 16  PaymentMethod     7032 n

In [7]:
# print categorical featrues summary
print ('\033[1m{:50s}\033[0m \n'.format("Categorical Features Description"))
for c in telco.columns:
    if c in cate_col:
        print ("---- %s ---" % c, "\n{}\n".format(telco[c].value_counts(normalize=True)))

# print numerical featrues summary
print ('\033[1m{:50s}\033[0m \n'.format("Numerical Features Description"))
for c in telco.columns:
    if c not in cate_col:
        print ("---- %s ---" % c, "\n{}\n".format(telco[c].describe()))    

Categorical Features Description                   

---- gender --- 
Male      0.504693
Female    0.495307
Name: gender, dtype: float64

---- SeniorCitizen --- 
No     0.8376
Yes    0.1624
Name: SeniorCitizen, dtype: float64

---- Partner --- 
No     0.517491
Yes    0.482509
Name: Partner, dtype: float64

---- Dependents --- 
No     0.701507
Yes    0.298493
Name: Dependents, dtype: float64

---- PhoneService --- 
Yes    0.903299
No     0.096701
Name: PhoneService, dtype: float64

---- MultipleLines --- 
No                  0.481371
Yes                 0.421928
No phone service    0.096701
Name: MultipleLines, dtype: float64

---- InternetService --- 
Fiber optic    0.440273
DSL            0.343572
No             0.216155
Name: InternetService, dtype: float64

---- OnlineSecurity --- 
No                     0.497298
Yes                    0.286547
No internet service    0.216155
Name: OnlineSecurity, dtype: float64

---- OnlineBackup --- 
No                     0.438993
Yes            

## Code Book

We organize the features info into the code book described below,

__CODE BOOK__


__Feature__      |__Type__|__Counts__|__Missing Value__|__Statistical Summary__|__Info Area__|__Description__| 
:-------------|:------|:------:|:------------:|:--------|:----|:-----------------------------------------|
Gender  |Category|7,032   |0% |Male: 54%<br>Female: 50%|Customer Demographic| Whether the customer is a male or a female| 
Senior Citizen |Category|7,032   |0% |Yes: 16%<br>No: 84%| Customer Demographic| Whether the customer is a senior citizen or not| 
Partner  |Category|7,032   |0% |Yes: 48%<br>No: 52%| Customer Demographic| Whether the customer has a partner or not|
Dependents|Category|7,032   |0% |Yes: 30%<br>No: 70%| Customer Demographic| Whether the customer has dependents or not|
Partner  |Category|7,032   |0% |Yes: 48%<br>No: 52%| Customer Demographic| Whether the customer has a partner or not|
Tenure|Numeric|7,032   |0% |Yes: 30%<br>No: 70%| Customer Account Information| Number of months the customer has stayed with the company|
Contract|Category|7,032   |0% |Month-to-month: 55%<br>One year: 21%<br>Two year: 24%| Customer Account Information| The contract term of the customer|
Paperless Billing|Category|7,032   |0% |Yes: 59%<br>No: 41%| Customer Account Information| Whether the customer has paperless billing or not|
Payment Method |Category|7,032   |0% |Electronic check: 33%<br>Mailed check: 23%<br>Bank transfer (automatic): 22%<br>Credit card (automatic): 21%| Customer Account Information| The customer’s payment method|
Monthly Charges|Numeric|7,032   |0% |Min: 18<br>25%:  36<br>50%: 70<br>75%: 90<br>Max: 119<br>Mean: 65|Customer account information|The amount charged to the customer monthly|
Total Charges |Numeric|7,032   |0% |Min: 19<br>25%:  401<br>50%: 1397<br>75%: 3994<br>Max: 8648<br>Mean: 2283|Customer account information| The total amount charged to the customer| 
Phone Service|Category|7,032   |0% |Yes: 90%<br>No: 10%| Customer Services| Whether the customer has a phone service or not|
Multiple Lines|Category|7,032   |0% |Yes: 42%<br>No: 48%<br>No phone service: 1%| Customer Services| Whether the customer has multiple lines or not |
Internet Service|Category|7,032   |0% |Fiber optic: 44%<br>DSL: 34%<br>No: 22%| Customer Services| Customer’s internet service provider |
Online Security|Category|7,032   |0% |Yes: 28%<br>No: 50%<br>No internet service: 22%| Customer Services| Whether the customer has online security or not|
Online Backup|Category|7,032   |0% |Yes: 34%<br>No: 44%<br>No internet service: 22%| Customer Services| Whether the customer has online backup or not|
Device Protection|Category|7,032   |0% |Yes: 34%<br>No: 44%<br>No internet service: 22% | Customer Services| Whether the customer has device protection or not|
Tech Support|Category|7,032   |0% |Yes: 29%<br>No: 49%<br>No internet service: 22% | Customer Services| Whether the customer has tech support or not|
Streaming TV|Category|7,032   |0% |Yes: 38%<br>No: 40%<br>No internet service: 22% | Customer Services| Whether the customer has streaming TV or not |
Streaming Movies|Category|7,032   |0% |Yes: 39%<br>No: 39%<br>No internet service: 22%| Customer Services| Whether the customer has streaming movies or not|

In [8]:
# save the cleaned data
telco.to_csv("telco_eda.csv",index=False)